api/src/main/java/org/eclipse/microprofile/reactive/messaging/Message.java

```diff
@@ -201,49 +210,179 @@ public Metadata getMetadata() {
+                return Metadata.empty();
+            }
+
+            @Override
+            public Supplier<CompletionStage<Void>> getAck() {
+                return ack;
+            }
+
+            @Override
+            public Function<Throwable, CompletionStage<Void>> getNack() {
+                return nack;
+            }
+        };
+    }
+
+    /**
+     * Create a message with the given payload, metadata and ack and nack functions.
+     *
+     * @param payload The payload.
+     * @param metadata the metadata, must not be {@code null}, must not contain {@code null} values.
+     * @param ack The ack function, this will be invoked when the returned messages {@link #ack()} method is invoked.
+     * @param nack The negative-ack function, this will be invoked when the returned messages {@link #nack(Throwable)}
+     *        method is invoked.
+     * @param <T> the type of payload
+     * @return A message with the given payload, metadata, ack and nack functions.
+     */
+    @Experimental("nack support is a SmallRye-only feature")
+    static <T> Message<T> of(T payload, Iterable<Object> metadata,
+            Supplier<CompletionStage<Void>> ack, Function<Throwable, CompletionStage<Void>> nack) {
+        Metadata validated = Metadata.from(metadata);
+        return new Message<T>() {
+            @Override
+            public T getPayload() {
+                return payload;
+            }
+
+            @Override
+            public Metadata getMetadata() {
+                return validated;
+            }
+
+            @Override
+            public Supplier<CompletionStage<Void>> getAck() {
+                return ack;
+            }
+
+            @Override
+            public Function<Throwable, CompletionStage<Void>> getNack() {
+                return nack;
+            }
+        };
+    }
+
+    /**
+     * Create a message with the given payload, metadata and ack and nack functions.
+     *
+     * @param payload The payload.
+     * @param metadata the metadata, must not be {@code null}, must not contain {@code null} values.
+     * @param ack The ack function, this will be invoked when the returned messages {@link #ack()} method is invoked.
+     * @param nack The negative-ack function, this will be invoked when the returned messages {@link #nack(Throwable)}
+     *        method is invoked.
+     * @param <T> the type of payload
+     * @return A message with the given payload, metadata, ack and nack functions.
+     */
+    @Experimental("nack support is a SmallRye-only feature")
+    static <T> Message<T> of(T payload, Metadata metadata,
+            Supplier<CompletionStage<Void>> ack, Function<Throwable, CompletionStage<Void>> nack) {
+        if (payload == null) {
+            throw new IllegalArgumentException("`payload` must not be `null`");
+        }
+        if (metadata == null) {
+            metadata = Metadata.empty();
+        }
+        Metadata actual = metadata;
```

Comment:
What's the purpose of this extra assignment?

Reply:
Because it must be (implicitly) final to be used in the inner class.